### PR TITLE
gather kube termination logs

### PR DESCRIPTION
**Pull request by rphillips**

### CODE REVIEW

File: collection-scripts/gather_apiserver_watchlog [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    sleep 2
+    oc wait -n "default" --for=condition=Ready pod/"$debugPod" --timeout=30s
+
+    if [ -z "$debugPod" ]
+    then
+      echo "Debug pod for node ""$1"" never activated"
+    else
+      #Copy API watch log
+      echo "Copying core dumps on node ""$1"""
+      oc cp  --loglevel 1 -n "default" "$debugPod":/host/var/log/kube-apiserver/termination.log "${WATCH_LOG_PATH}"/"$1"_termination.log > /dev/null 2>&1 && PIDS+=($!)
```
Comment (smarterclayton) [inline collection-scripts/gather_apiserver_watchlog:26]: If the file name is called "termination.log", the script and the function name output should be called  "kube-apiserver termination logs".  Watchlog doesn't make sense

Comment (rphillips) [reply collection-scripts/gather_apiserver_watchlog]: Updated. Thanks

File: collection-scripts/gather_kube_apiserver_termination_logs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+function gather_api_watchlog_data {
+  #Run watchlog pull function on all nodes in parallel
+  for NODE in ${NODES}; do
+    get_dump_off_node "${NODE}" &
+  done
+}
+
+if [ $# -eq 0 ]; then
+    echo "WARNING: Collecting watchlog dumps on ALL linux nodes in your cluster. This could take a long time."
```
Comment (smarterclayton) [inline collection-scripts/gather_kube_apiserver_termination_logs:41]: update all this text too, I have no idea what a watchlog is (and this is all output someone will have to use to find this script)

Comment (rphillips) [reply collection-scripts/gather_kube_apiserver_termination_logs]: Updated to

```
WARNING: Collecting kube apiserver watch termination logs on ALL linux nodes in your cluster.
```

File: collection-scripts/gather_kube_apiserver_termination_logs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    sleep 2
+    oc wait -n "default" --for=condition=Ready pod/"$debugPod" --timeout=30s
+
+    if [ -z "$debugPod" ]
+    then
+      echo "Debug pod for node ""$1"" never activated"
+    else
+      #Copy API watch log
+      echo "Copying termination log on node ""$1"""
+      oc cp  --loglevel 1 -n "default" "$debugPod":/host/var/log/kube-apiserver/termination.log "${TERMINATION_LOG_PATH}"/"$1"_termination.log > /dev/null 2>&1 && PIDS+=($!)
```
Comment (deads2k) [inline collection-scripts/gather_kube_apiserver_termination_logs:26]: Is this attempting to do https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_audit_logs#L16 ?

Comment (rphillips) [reply collection-scripts/gather_kube_apiserver_termination_logs:26]: This PR isn't gathering the audit logs. It is gathering the watch termination logs from the API server.

Comment (deads2k) [reply collection-scripts/gather_kube_apiserver_termination_logs:26]: Why wouldn't `oc adm node-logs --role=master --path="$path"` do the job?

File: collection-scripts/gather_kube_apiserver_termination_logs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+# Downloads the apiserver watch termination logs (and its rotated copies) from
+# /var/logs/kube-apiserver/termination.log on each master node.
+BASE_COLLECTION_PATH="${BASE_COLLECTION_PATH:-/must-gather}"
+
+oc adm node-logs --role=master --path="kube-apiserver/termination.log" > ${BASE_COLLECTION_PATH}/apiserver_watch_termination.log
```
Comment (deads2k) [inline collection-scripts/gather_kube_apiserver_termination_logs:6]: question: this properly separates the three files?

**Comment by deads2k** [comment]
/lgtm
/hold

hold for proof of collection.

File: collection-scripts/gather_kube_apiserver_termination_logs (hunk 4 above)
Comment (rphillips) [reply collection-scripts/gather_kube_apiserver_termination_logs:6]: it stores them into one file.

Comment (rphillips) [reply collection-scripts/gather_kube_apiserver_termination_logs:6]: I'll tweak it to split them up.

**Comment by rphillips** [comment]
Audit logs were disabled here: https://github.com/openshift/must-gather/pull/144/files

This re-enables them.

File: collection-scripts/gather_kube_apiserver_termination_logs (hunk 4 above)
Comment (rphillips) [reply collection-scripts/gather_kube_apiserver_termination_logs]: for some reason gather_audit_logs were disabled... I have re-enabled them for must gather and it does include the termination.log

File: collection-scripts/gather_kubeapi_termination_logs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+# Downloads the termination.log (and its rotated copies) from
+# /var/logs/{kube-apiserver,openshift-apiserver} on each master node.
+BASE_COLLECTION_PATH="${BASE_COLLECTION_PATH:-/must-gather}"
+echo "WARNING: Collecting one or more watch termination logs on ALL masters in your cluster." >&2
```
Comment (sferich888) [inline collection-scripts/gather_kubeapi_termination_logs:5]: I don't know that this is needed (as a warning). Can we switch this to INFO?

File: collection-scripts/gather_kubeapi_termination_logs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  output_dir="${BASE_COLLECTION_PATH}/kube-watch-termination-logs/$path"
+  mkdir -p "$output_dir"
+  oc adm node-logs --role=master --path="$path" | \
+  tee "${BASE_COLLECTION_PATH}/kube-watch-termination-logs/$path.watch_logs_listing" | \
+  grep "termination.log" | \
+  sed "s|^|$path $output_dir |"
+done | \
+xargs --max-args=4 --max-procs=45 bash -c \
+  'echo "INFO: Started downloading $1/$4 from $3";
+  oc adm node-logs $3 --path=$1/$4 | gzip > $2/$3-$4.gz;
```
Comment (sferich888) [inline collection-scripts/gather_kubeapi_termination_logs:22]: Should we extend the utility function in https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_service_logs_util to do this? 

Why are we gziping it at this stage? Is this a lot of data? If so we might have a concern with collecting this. 

https://issues.redhat.com/browse/RFE-1576 should take care of compression of the full archive post collection.

**Review by sferich888 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**